Research/lva-node-app/public/js/main.js

```diff
@@ -0,0 +1,667 @@
+      break;
+  }
+
+  checkBoxNameList.forEach((item) => 
+  {
+    hoverList.innerHTML += "<li> <button class='btn btn-outline-secondary' onclick='"+onclickMethod+"' name='" + methodName +"'>" + item + "</button> </li>";
+  });
+}
+
+/**
+* Create the larger payload object to be sent to the server to invoke a direct method on the lvaEdge module
+* @param {string} methodNameParam - name of method to invoke
+* @param {jsonObject} methodPayload - payload for method
+* @returns {jsonObject} full payload for request
+*/
+function createFullPayload(methodNameParam, methodPayload)
+{
+  return {
+    methodName: methodNameParam,
+    Payload: methodPayload  
+  }
+}
+
+/**
+* see sample graph topologies here: https://github.com/Azure/live-video-analytics/tree/master/MediaGraph/topologies
+* @param {string} topologyName - graph topology name
+* @returns {void}
+*/
+function graphSetTopology(topologyName)
+{
+  let topology = graphTopologies[topologyName];
+  if(topology == undefined)
+  {
+    alert("Topology "+topologyName+" not found.");
+    return;
+  }
+  invokeLVAMethod(createFullPayload("GraphTopologySet", topology)).then((response) =>
+  {
+    // do nothing, invokeLVAMethod handles
+  }).catch((error) =>
+  {
+    console.error(error);
+  });
+}
+
+/**
+* invoke a list method on lvaEdge module (either list topology or list instances. payload is the same)
+* @param {string} methodName - method name (either GraphTopologyList or GraphInstanceList)
+* @returns {Promise<any>}
+*/
+function graphEntityList(methodName) 
+{
+  return new Promise((resolve, reject) =>
+  {
+    let payload = 
+    {
+      "@apiVersion": "1.0"
+    }
+    
+    invokeLVAMethod(createFullPayload(methodName, payload)).then((response) =>
+    {
+      resolve(response);
+    }).catch((error) =>
+    {
+      reject(error);
+    });
+  });
+}
+
+/**
+* this payload is used for GraphInstanceDelete, GraphTopologyDelete, GraphInstanceActivate, GraphInstanceDeactivate
+* @param {string | HTMLButtonElement} htmlElement - button clicked on to invoke method on given element or string of element name (i.e. "Graph-Instance-1")
+* @param {string} methodName - method name
+* @param {boolean} elementNamePassed - true if function caller passed in the 
+*/
+function graphEntityModify(htmlElement, methodName, elementNamePassed=false) 
+{
+  let elementName = elementNamePassed ? htmlElement : htmlElement.innerText;
+  let payload = 
+  {
+    "@apiVersion": "1.0",
+    name: elementName
+  }
+  invokeLVAMethod(createFullPayload(methodName, payload)).then((response) =>
+  {
+    // do nothing, invokeLVAMethod handles
+  }).catch((error) =>
+  {
+    console.error(error);
+  });;
+}
+
+
+/**
+* calls GraphInstanceList and GraphTopologyList. Runs on load of output.html
+* @async
+* @returns {Promise<void>}
+*/
+async function loadInstancesAndTopologies()
+{
+  getGlobals().then((response) =>
+  {
+    graphEntityList("GraphInstanceList").then(() =>
```

Comment:
i'm clear on what getGlobals does, but not clear on the rest of this... it doesn't look like we use the response from getGlobals in the below, nor do I see the output flowing anywhere

Reply:
we are pulling in the cameras and any topologies the user has already set. Then with the service we don't overwrite - we just add/modify. for example GraphInstanceList would have never been called if this is the user's first time visiting the output page, so that will modify graphInstances. If the user never visited the mediagraphs page, then it would also be running GraphTopologyList for the first time

Reply:
got rid of graphInstances and graphTopologies as global variables. This also resolves a potential security vulnerability regarding cameras that were previously set up